File: src/lib/wow/archive/casc/listfile.ts
/**
 * Master listfile management, ported from wow.export (src/js/casc/listfile.js).
 * UI tab pre-filtering (textures/sounds/videos/text/models arrays) is omitted;
 * unknown-file providers are injected by the DB cache modules (Phase 4).
 */
import { promises as fsp } from 'fs';
import path from 'path';

import { write } from '@/lib/wow/log';

import { BufferWrapper } from '../../formats/buffer';
import { constants } from '../../formats/constants';
import { batchWork, downloadFile } from '../../formats/generics';
import { wowConfig } from '../../server/config';

const nameLookup = new Map<string, number>();
const idLookup = new Map<number, string>();

let loaded = false;

const preloadedIdLookup = new Map<number, string>();
const preloadedNameLookup = new Map<string, number>();
let isPreloaded = false;
let preloadPromise: Promise<boolean> | null = null;

/** Replace the extension of a file path (replaceExtension equivalent). */
function replaceExtension(file: string, ext = ''): string {
  return path.join(path.dirname(file), path.basename(file, path.extname(file)) + ext);
}

/** Internal implementation of preload logic. */
async function _doPreload(): Promise<boolean> {
  try {
    write('Preloading master listfile...');

    const url = String(wowConfig.listfileURL);
    if (typeof url !== 'string') throw new Error('Missing/malformed listfileURL in configuration!');

    // Ensure listfile cache directory exists
    await fsp.mkdir(constants.CACHE.DIR_LISTFILE, { recursive: true });

    const cacheFile = path.join(constants.CACHE.DIR_LISTFILE, constants.CACHE.LISTFILE_DATA);

    preloadedIdLookup.clear();
    preloadedNameLookup.clear();
    isPreloaded = false;

    let data: BufferWrapper;
    if (url.startsWith('http')) {
      // Listfile URL is http, check for cache/updates
      let requireDownload = false;
      let cached: BufferWrapper | null = null;
      let lastModified = 0;

      // Try to load existing cached data and get its modification time
      try {
        cached = await BufferWrapper.readFile(cacheFile);
        const stats = await fsp.stat(cacheFile);
        lastModified = stats.mtime.getTime();
      } catch (e) {
        // No cached file
      }

      if (lastModified > 0) {
        let ttl = Number(wowConfig.listfileCacheRefresh) || 0;
        ttl *= 24 * 60 * 60 * 1000; // Reduce from days to milliseconds.

        if (ttl === 0 || (Date.now() - lastModified) > ttl) {
          // Local cache file needs updating.
          write('Cached listfile is out-of-date (> %d).', ttl);
          requireDownload = true;
        } else if (cached === null) {
          // Ensure that the local cache file *actually* exists before relying on it.
          write('Listfile is missing despite file stats. User tamper?');
          requireDownload = true;
        } else {
          write('Listfile is cached locally.');
        }
      } else {
        // This listfile has never been cached.
        requireDownload = true;
        write('Listfile is not cached, downloading fresh.');
      }

      if (requireDownload) {
        try {
          let fallbackUrl = String(wowConfig.listfileFallbackURL);
          // Remove %s placeholder since we don't use buildConfig for master listfile
          fallbackUrl = fallbackUrl.replace('%s', '');

          data = await downloadFile([url, fallbackUrl]);

          // Store the downloaded data (file modification time will be set automatically)
          await fsp.writeFile(cacheFile, data.raw);
        } catch (e) {
          if (cached === null) {
            write('Failed to download listfile during preload, no cached version for fallback: %s', (e as Error).message);
            return false;
          }

          write('Failed to download listfile during preload, using cached version: %s', (e as Error).message);
          data = cached;
        }
      } else {
        data = cached!;
      }
    } else {
      // User has configured a local listfile location
      write('Preloading user-defined local listfile: %s', url);
      data = await BufferWrapper.readFile(url);
    }

    const lines = data.readLines();
    write('Processing %d listfile lines in chunks...', lines.length);

    await batchWork('listfile parsing', lines, (line) => {
      if (line.length === 0) return;

      const tokens = line.split(';');
      if (tokens.length !== 2) {
        write('Invalid listfile line (token count): %s', line);
        return;
      }

      const fileDataID = Number(tokens[0]);
      if (Number.isNaN(fileDataID)) {
        write('Invalid listfile line (non-numerical ID): %s', line);
        return;
      }

      const fileName = tokens[1].toLowerCase();
      preloadedIdLookup.set(fileDataID, fileName);
      preloadedNameLookup.set(fileName, fileDataID);
    }, 50000);

    if (preloadedIdLookup.size === 0) {
      write('No entries found in preloaded listfile');
      return false;
    }

    isPreloaded = true;
    write('Preloaded %d listfile entries', preloadedIdLookup.size);
    return true;
  } catch (e) {
    write('Error during listfile preload: %s', (e as Error).message);
    isPreloaded = false;
    return false;
  }
}

/**
 * Preload the master listfile.
 * Multiple calls to this function will return the same promise.
 */
export async function preload(): Promise<boolean> {
  if (preloadPromise) return preloadPromise;
  if (isPreloaded) return true;

  preloadPromise = _doPreload();
  return preloadPromise;
}

/**
 * Ensure listfile is preloaded and ready for use.
 * This should be called during the loading process before accessing listfile data.
 */
export async function prepareListfile(): Promise<boolean> {
  if (isPreloaded) return true;

  if (preloadPromise) {
    write('Waiting for listfile preload to complete...');
    return preloadPromise;
  }

  write('Starting listfile preload...');
  return preload();
}

/**
 * Apply preloaded listfile data filtered by rootEntries.
 * @returns Number of entries applied, or 0 if preload not available/failed.
 */
export function applyPreload(rootEntries: Map<number, unknown>): number {
  if (!isPreloaded) {
    write('No preloaded listfile available, falling back to normal loading');
    return 0;
  }

  try {
    write('Applying preloaded listfile data...');

    // Clear current data
    idLookup.clear();
    nameLookup.clear();

    // Apply preloaded entries filtered by rootEntries
    let appliedCount = 0;
    for (const [fileDataID, fileName] of preloadedIdLookup.entries()) {
      if (rootEntries.has(fileDataID)) {
        idLookup.set(fileDataID, fileName);
        nameLookup.set(fileName, fileDataID);
        appliedCount++;
      }
    }

    if (appliedCount === 0) {
      write('No preloaded entries matched rootEntries');
      return 0;
    }

    loaded = true;
    write('Applied %d preloaded listfile entries', appliedCount);

    // Active idLookup/nameLookup now hold the filtered subset; drop the full preload
    // (~2M entries) to reclaim RAM. A later CASC build switch re-parses from disk cache.
    preloadedIdLookup.clear();
    preloadedNameLookup.clear();
    isPreloaded = false;
    preloadPromise = null;

    return appliedCount;
  } catch (e) {
    write('Error applying preloaded listfile: %s', (e as Error).message);
    return 0;
  }
}

/** Providers of unknown fileDataIDs, registered by DB cache modules. */
type UnknownIdProvider = () => Iterable<number> | Promise<Iterable<number>>;
let unknownModelProvider: UnknownIdProvider | null = null;
let unknownTextureProvider: UnknownIdProvider | null = null;

export function setUnknownModelProvider(provider: UnknownIdProvider): void {
  unknownModelProvider = provider;
}

export function setUnknownTextureProvider(provider: UnknownIdProvider): void {
  unknownTextureProvider = provider;
}

/** Load unknown texture files from TextureFileData. */
export async function loadUnknownTextures(): Promise<number> {
  if (!unknownTextureProvider) return 0;
  const unkBlp = loadIDTable(await unknownTextureProvider(), '.blp');
  write('Added %d unknown BLP textures from TextureFileData to listfile', unkBlp);
  return unkBlp;
}

/** Load unknown model files from ModelFileData. */
export async function loadUnknownModels(): Promise<number> {
  if (!unknownModelProvider) return 0;
  const unkM2 = loadIDTable(await unknownModelProvider(), '.m2');
  write('Added %d unknown M2 models from ModelFileData to listfile', unkM2);
  return unkM2;
}

/**
 * Load unknown files from ModelFileData only.
 * TextureFileData unknown files are loaded separately.
 */
export async function loadUnknowns(): Promise<void> {
  await loadUnknownModels();
}

/** Load file IDs from a data table. */
export function loadIDTable(ids: Iterable<number>, ext: string): number {
  let loadCount = 0;

  for (const fileDataID of ids) {
    if (!idLookup.has(fileDataID)) {
      const fileName = `unknown/${fileDataID}${ext}`;
      idLookup.set(fileDataID, fileName);
      nameLookup.set(fileName, fileDataID);
      loadCount++;
    }
  }

  return loadCount;
}

export type ExtensionFilter = string | [string, RegExp];

/** Return an array of filenames ending with the given extension(s). */
export function getFilenamesByExtension(exts: ExtensionFilter | ExtensionFilter[]): string[] {
  // Box into an array for reduced code.
  let filters = exts;
  if (!Array.isArray(exts) || (exts.length === 2 && typeof exts[0] === 'string' && exts[1] instanceof RegExp)) {
    filters = [exts as ExtensionFilter];
  }

  const entries: number[] = [];

  for (const [fileDataID, filename] of idLookup.entries()) {
    for (const ext of filters as ExtensionFilter[]) {
      if (Array.isArray(ext)) {
        if (filename.endsWith(ext[0]) && !filename.match(ext[1])) {
          entries.push(fileDataID);
          continue;
        }
      } else if (filename.endsWith(ext)) {
        entries.push(fileDataID);
        continue;
      }
    }
  }

  return formatEntries(entries);
}

/** Sort and format listfile entries for file list display. */
export function formatEntries(entries: number[]): string[] {
  // If sorting by ID, perform the sort while the array is only IDs.
  if (wowConfig.listfileSortByID) entries.sort((a, b) => a - b);

  let formatted: string[];
  if (wowConfig.listfileShowFileDataIDs) formatted = entries.map((e) => `${getByIDOrUnknown(e)} [${e}]`);
  else formatted = entries.map((e) => getByIDOrUnknown(e));

  // If sorting by name, sort now that the filenames have been added.
  if (!wowConfig.listfileSortByID) formatted.sort();

  return formatted;
}

export function ingestIdentifiedFiles(entries: Map<number, string>): void {
  for (const [fileDataID, ext] of entries) {
    const fileName = `unknown/${fileDataID}${ext}`;
    idLookup.set(fileDataID, fileName);
    nameLookup.set(fileName, fileDataID);
  }
}

/** Returns a full listfile, sorted and formatted. */
export function getFullListfile(): string[] {
  return formatEntries([...idLookup.keys()]);
}

/** Get a filename from a given file data ID. */
export function getByID(id: number): string | undefined {
  return idLookup.get(id);
}

/** Get a filename from a given file data ID or format it as an unknown file. */
export function getByIDOrUnknown(id: number, ext = ''): string {
  return idLookup.get(id) ?? formatUnknownFile(id, ext);
}

/** Get a file data ID by a given file name. */
export function getByFilename(filename: string): number | undefined {
  let lookup = nameLookup.get(filename.toLowerCase().replace(/\\/g, '/'));

  // In the rare occasion we have a reference to an MDL/MDX file and it fails
  // to resolve (as expected), attempt to resolve the M2 of the same name.
  if (!lookup && (filename.endsWith('.mdl') || filename.endsWith('mdx'))) {
    lookup = nameLookup.get(replaceExtension(filename, '.m2').replace(/\\/g, '/'));
  }

  return lookup;
}

export interface ListfileEntry {
  fileDataID: number;
  fileName: string;
}

/** Returns an array of listfile entries filtered by the given search term. */
export function getFilteredEntries(search: string | RegExp): ListfileEntry[] {
  const results: ListfileEntry[] = [];
  const isRegExp = search instanceof RegExp;

  for (const [fileDataID, fileName] of idLookup.entries()) {
    if (search === '' || (isRegExp ? fileName.match(search) : fileName.includes(search as string))) {
      results.push({ fileDataID, fileName });
    }
  }

  return results;
}

/** Strips a prefixed file ID from a listfile entry. */
export function stripFileEntry(entry: string): string {
  if (typeof entry === 'string' && entry.includes(' [')) return entry.substring(0, entry.lastIndexOf(' ['));

  return entry;
}

/** Returns a file path for an unknown fileDataID. */
export function formatUnknownFile(fileDataID: number, ext = ''): string {
  return `unknown/${fileDataID}${ext}`;
}

/** Returns true if a listfile has been loaded. */
export function isLoaded(): boolean {
  return loaded;
}

export function getMemoryStats(): {
  idLookup: number;
  nameLookup: number;
  preloadedIdLookup: number;
  preloadedNameLookup: number;
  isPreloaded: boolean;
  loaded: boolean;
  } {
  return {
    idLookup: idLookup.size,
    nameLookup: nameLookup.size,
    preloadedIdLookup: preloadedIdLookup.size,
    preloadedNameLookup: preloadedNameLookup.size,
    isPreloaded,
    loaded,
  };
}

/** Adds an entry to the listfile. */
export function addEntry(fileDataID: number, fileName: string): void {
  idLookup.set(fileDataID, fileName);
  nameLookup.set(fileName, fileDataID);
}

/** Clear in-memory listfile state so a new CASC build can load cleanly. */
export function resetForCascUnload(): void {
  loaded = false;
  isPreloaded = false;
  preloadPromise = null;
  idLookup.clear();
  nameLookup.clear();
  preloadedIdLookup.clear();
  preloadedNameLookup.clear();
  browseModels = null;
  browseTextures = null;
}

const browseM2WmoRegex = /\.(m2|wmo)$/i;
const browseBadWmoRegex = /_([0-9]{3}|lod\d)\.wmo$/i;
const browseTextureRegex = /\.(blp|png|tga|dds)$/i;

let browseModels: ListfileEntry[] | null = null;
let browseTextures: ListfileEntry[] | null = null;

function buildBrowseFileIndex(): { models: ListfileEntry[]; textures: ListfileEntry[] } {
  const models: ListfileEntry[] = [];
  const textures: ListfileEntry[] = [];
  for (const [fileDataID, fileName] of idLookup.entries()) {
    if (browseBadWmoRegex.test(fileName)) continue;
    if (browseM2WmoRegex.test(fileName)) {
      models.push({ fileDataID, fileName });
    } else if (browseTextureRegex.test(fileName)) {
      textures.push({ fileDataID, fileName });
    }
  }
  models.sort((a, b) => a.fileName.localeCompare(b.fileName));
  textures.sort((a, b) => a.fileName.localeCompare(b.fileName));
  return { models, textures };
}

/** Returns cached browse model/texture indexes (mirrors Go CollectBrowseFileIndex). */
export function collectBrowseFileIndex(): { models: ListfileEntry[]; textures: ListfileEntry[] } {
  if (browseModels !== null && browseTextures !== null) {
    return { models: browseModels, textures: browseTextures };
  }
  const built = buildBrowseFileIndex();
  browseModels = built.models;
  browseTextures = built.textures;
  return built;
}

export default {
  loadUnknowns,
  loadUnknownTextures,
  loadUnknownModels,
  loadIDTable,
  preload,
  prepareListfile,
  applyPreload,
  getByID,
  getByFilename,
  getFullListfile,
  getFilenamesByExtension,
  getFilteredEntries,
  getByIDOrUnknown,
  stripFileEntry,
  formatEntries,
  formatUnknownFile,
  ingestIdentifiedFiles,
  isLoaded,
  addEntry,
  collectBrowseFileIndex,
  setUnknownModelProvider,
  setUnknownTextureProvider,
};

File: golang/internal/wow/archive/casc/listfile.go
package casc

import (
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/pqhuy98/wow-converter/internal/buffer"
	"github.com/pqhuy98/wow-converter/internal/stringsort"
	"github.com/pqhuy98/wow-converter/internal/wow/server"
	"github.com/pqhuy98/wow-converter/internal/wow/constants"
	"github.com/pqhuy98/wow-converter/internal/wow/formats"
	"github.com/pqhuy98/wow-converter/internal/wow/log"
)

var (
	nameLookup          = map[string]int{}
	idLookup            = map[int]string{}
	listfileLoaded      bool
	preloadedIDLookup   = map[int]string{}
	preloadedNameLookup = map[string]int{}
	isPreloaded         bool
	preloadMu           sync.Mutex
	preloadPromise      chan preloadResult
)

type preloadResult struct {
	ok  bool
	err error
}

type unknownIDProvider func() ([]int, error)

var (
	unknownModelProvider   unknownIDProvider
	unknownTextureProvider unknownIDProvider
)

func replaceExtension(file, ext string) string {
	return filepath.Join(filepath.Dir(file), strings.TrimSuffix(filepath.Base(file), filepath.Ext(file))+ext)
}

func doPreload() (bool, error) {
	log.Write("Preloading master listfile...")
	url := server.GetConfig().ListfileURL
	if url == "" {
		return false, errMissingListfileURL
	}
	if err := os.MkdirAll(constants.Cache.DirListfile, 0o755); err != nil {
		return false, err
	}
	cacheFile := filepath.Join(constants.Cache.DirListfile, constants.Cache.ListfileData)
	preloadedIDLookup = map[int]string{}
	preloadedNameLookup = map[string]int{}
	isPreloaded = false

	var data *buffer.Buffer
	if strings.HasPrefix(url, "http") {
		requireDownload := false
		var cached *buffer.Buffer
		lastModified := int64(0)
		if c, err := buffer.ReadFile(cacheFile); err == nil {
			cached = c
			if st, err := os.Stat(cacheFile); err == nil {
				lastModified = st.ModTime().UnixMilli()
			}
		}
		if lastModified > 0 {
			ttl := int64(server.GetConfig().ListfileCacheRefresh) * 24 * 60 * 60 * 1000
			if ttl == 0 || (time.Now().UnixMilli()-lastModified) > ttl {
				log.Write("Cached listfile is out-of-date (> %d).", ttl)
				requireDownload = true
			} else if cached == nil {
				log.Write("Listfile is missing despite file stats. User tamper?")
				requireDownload = true
			} else {
				log.Write("Listfile is cached locally.")
			}
		} else {
			requireDownload = true
			log.Write("Listfile is not cached, downloading fresh.")
		}
		if requireDownload {
			fallbackURL := strings.ReplaceAll(server.GetConfig().ListfileFallbackURL, "%s", "")
			downloaded, err := formats.DownloadFile([]string{url, fallbackURL}, "", -1, -1, false)
			if err != nil {
				if cached == nil {
					log.Write("Failed to download listfile during preload, no cached version for fallback: %s", err.Error())
					return false, nil
				}
				log.Write("Failed to download listfile during preload, using cached version: %s", err.Error())
				data = cached
			} else {
				data = downloaded
				_ = os.WriteFile(cacheFile, data.Raw(), 0o644)
			}
		} else {
			data = cached
		}
	} else {
		log.Write("Preloading user-defined local listfile: %s", url)
		var err error
		data, err = buffer.ReadFile(url)
		if err != nil {
			return false, err
		}
	}
	lines := data.ReadLines("utf8")
	log.Write("Processing %d listfile lines...", len(lines))
	for _, line := range lines {
		if line == "" {
			continue
		}
		tokens := strings.Split(line, ";")
		if len(tokens) != 2 {
			log.Write("Invalid listfile line (token count): %s", line)
			continue
		}
		fileDataID, err := strconv.Atoi(tokens[0])
		if err != nil {
			log.Write("Invalid listfile line (non-numerical ID): %s", line)
			continue
		}
		fileName := strings.ToLower(tokens[1])
		preloadedIDLookup[fileDataID] = fileName
		preloadedNameLookup[fileName] = fileDataID
	}
	if len(preloadedIDLookup) == 0 {
		log.Write("No entries found in preloaded listfile")
		return false, nil
	}
	isPreloaded = true
	log.Write("Preloaded %d listfile entries", len(preloadedIDLookup))
	return true, nil
}

// Preload preloads the master listfile.
func Preload() (bool, error) {
	preloadMu.Lock()
	defer preloadMu.Unlock()
	if preloadPromise != nil {
		res := <-preloadPromise
		return res.ok, res.err
	}
	if isPreloaded {
		return true, nil
	}
	ch := make(chan preloadResult, 1)
	preloadPromise = ch
	go func() {
		ok, err := doPreload()
		ch <- preloadResult{ok: ok, err: err}
	}()
	res := <-ch
	preloadPromise = nil
	return res.ok, res.err
}

// PrepareListfile ensures listfile is preloaded.
func PrepareListfile() (bool, error) {
	if isPreloaded {
		return true, nil
	}
	preloadMu.Lock()
	if preloadPromise != nil {
		ch := preloadPromise
		preloadMu.Unlock()
		log.Write("Waiting for listfile preload to complete...")
		res := <-ch
		return res.ok, res.err
	}
	preloadMu.Unlock()
	log.Write("Starting listfile preload...")
	return Preload()
}

// ApplyPreload applies preloaded listfile data filtered by rootEntries.
func ApplyPreload(rootEntries map[int]RootEntry) int {
	if !isPreloaded {
		log.Write("No preloaded listfile available, falling back to normal loading")
		return 0
	}
	log.Write("Applying preloaded listfile data...")
	idLookup = map[int]string{}
	nameLookup = map[string]int{}
	appliedCount := 0
	for fileDataID, fileName := range preloadedIDLookup {
		if _, ok := rootEntries[fileDataID]; ok {
			idLookup[fileDataID] = fileName
			nameLookup[fileName] = fileDataID
			appliedCount++
		}
	}
	if appliedCount == 0 {
		log.Write("No preloaded entries matched rootEntries")
		return 0
	}
	listfileLoaded = true
	log.Write("Applied %d preloaded listfile entries", appliedCount)
	preloadedIDLookup = map[int]string{}
	preloadedNameLookup = map[string]int{}
	isPreloaded = false
	preloadPromise = nil
	return appliedCount
}

// SetUnknownModelProvider registers a provider of unknown model fileDataIDs.
func SetUnknownModelProvider(provider unknownIDProvider) {
	unknownModelProvider = provider
}

// SetUnknownTextureProvider registers a provider of unknown texture fileDataIDs.
func SetUnknownTextureProvider(provider unknownIDProvider) {
	unknownTextureProvider = provider
}

// LoadUnknownTextures loads unknown texture files.
func LoadUnknownTextures() (int, error) {
	if unknownTextureProvider == nil {
		return 0, nil
	}
	ids, err := unknownTextureProvider()
	if err != nil {
		return 0, err
	}
	count := LoadIDTable(ids, ".blp")
	log.Write("Added %d unknown BLP textures from TextureFileData to listfile", count)
	return count, nil
}

// LoadUnknownModels loads unknown model files.
func LoadUnknownModels() (int, error) {
	if unknownModelProvider == nil {
		return 0, nil
	}
	ids, err := unknownModelProvider()
	if err != nil {
		return 0, err
	}
	count := LoadIDTable(ids, ".m2")
	log.Write("Added %d unknown M2 models from ModelFileData to listfile", count)
	return count, nil
}

// LoadUnknowns loads unknown files from ModelFileData.
func LoadUnknowns() error {
	_, err := LoadUnknownModels()
	return err
}

// LoadIDTable loads file IDs from a data table.
func LoadIDTable(ids []int, ext string) int {
	loadCount := 0
	for _, fileDataID := range ids {
		if _, ok := idLookup[fileDataID]; !ok {
			fileName := FormatUnknownFile(fileDataID, ext)
			idLookup[fileDataID] = fileName
			nameLookup[fileName] = fileDataID
			loadCount++
		}
	}
	return loadCount
}

// ExtensionFilter filters listfile entries by extension.
type ExtensionFilter any

// GetFilenamesByExtension returns filenames ending with the given extension(s).
func GetFilenamesByExtension(exts ExtensionFilter) []string {
	filters := []ExtensionFilter{exts}
	if arr, ok := exts.([]ExtensionFilter); ok {
		filters = arr
	}
	var entries []int
	for fileDataID, filename := range idLookup {
		for _, ext := range filters {
			switch e := ext.(type) {
			case [2]any:
				suffix, _ := e[0].(string)
				re, _ := e[1].(*regexp.Regexp)
				if strings.HasSuffix(filename, suffix) && (re == nil || !re.MatchString(filename)) {
					entries = append(entries, fileDataID)
				}
			case string:
				if strings.HasSuffix(filename, e) {
					entries = append(entries, fileDataID)
				}
			}
		}
	}
	return FormatEntries(entries)
}

// FormatEntries sorts and formats listfile entries.
func FormatEntries(entries []int) []string {
	cfg := server.GetConfig()
	if cfg.ListfileSortByID {
		sort.Ints(entries)
	}
	formatted := make([]string, len(entries))
	for i, e := range entries {
		if cfg.ListfileShowFileDataIDs {
			formatted[i] = GetByIDOrUnknown(e, "") + " [" + strconv.Itoa(e) + "]"
		} else {
			formatted[i] = GetByIDOrUnknown(e, "")
		}
	}
	if !cfg.ListfileSortByID {
		sort.Strings(formatted)
	}
	return formatted
}

// IngestIdentifiedFiles adds identified unknown files.
func IngestIdentifiedFiles(entries map[int]string) {
	for fileDataID, ext := range entries {
		fileName := FormatUnknownFile(fileDataID, ext)
		idLookup[fileDataID] = fileName
		nameLookup[fileName] = fileDataID
	}
}

// GetFullListfile returns a full listfile.
func GetFullListfile() []string {
	keys := make([]int, 0, len(idLookup))
	for k := range idLookup {
		keys = append(keys, k)
	}
	return FormatEntries(keys)
}

// GetByID returns a filename for a file data ID.
func GetByID(id int) (string, bool) {
	name, ok := idLookup[id]
	return name, ok
}

// GetByIDOrUnknown returns a filename or unknown path.
func GetByIDOrUnknown(id int, ext string) string {
	if name, ok := idLookup[id]; ok {
		return name
	}
	return FormatUnknownFile(id, ext)
}

// GetByFilename returns a file data ID by filename.
func GetByFilename(filename string) (int, bool) {
	lookup, ok := nameLookup[strings.ToLower(strings.ReplaceAll(filename, "\\", "/"))]
	if !ok && (strings.HasSuffix(filename, ".mdl") || strings.HasSuffix(filename, "mdx")) {
		alt := strings.ReplaceAll(replaceExtension(filename, ".m2"), "\\", "/")
		lookup, ok = nameLookup[strings.ToLower(alt)]
	}
	return lookup, ok
}

// ListfileEntry is a filtered listfile entry.
type ListfileEntry struct {
	FileDataID int
	FileName   string
}

// GetFilteredEntries returns listfile entries matching search.
func GetFilteredEntries(search any) []ListfileEntry {
	var results []ListfileEntry
	isRegExp := false
	var re *regexp.Regexp
	var term string
	switch s := search.(type) {
	case *regexp.Regexp:
		isRegExp = true
		re = s
	case string:
		term = s
	}
	for fileDataID, fileName := range idLookup {
		match := false
		if isRegExp {
			match = re.MatchString(fileName)
		} else {
			match = term == "" || strings.Contains(fileName, term)
		}
		if match {
			results = append(results, ListfileEntry{FileDataID: fileDataID, FileName: fileName})
		}
	}
	return results
}

// StripFileEntry strips a prefixed file ID from a listfile entry.
func StripFileEntry(entry string) string {
	if strings.Contains(entry, " [") {
		return entry[:strings.LastIndex(entry, " [")]
	}
	return entry
}

// FormatUnknownFile returns a file path for an unknown fileDataID.
func FormatUnknownFile(fileDataID int, ext string) string {
	return "unknown/" + strconv.Itoa(fileDataID) + ext
}

// IsListfileLoaded reports whether a listfile has been loaded.
func IsListfileLoaded() bool {
	return listfileLoaded
}

// ListfileMemoryStats reports in-memory listfile sizes.
type ListfileMemoryStats struct {
	IDLookup            int
	NameLookup          int
	PreloadedIDLookup   int
	PreloadedNameLookup int
	IsPreloaded         bool
	Loaded              bool
}

// GetMemoryStats returns listfile memory stats.
func GetMemoryStats() ListfileMemoryStats {
	return ListfileMemoryStats{
		IDLookup:            len(idLookup),
		NameLookup:          len(nameLookup),
		PreloadedIDLookup:   len(preloadedIDLookup),
		PreloadedNameLookup: len(preloadedNameLookup),
		IsPreloaded:         isPreloaded,
		Loaded:              listfileLoaded,
	}
}

// IndexMemoryStats reports built browse/map tile indexes held in RAM.
type IndexMemoryStats struct {
	BrowseBuilt      bool
	BrowseModels     int
	BrowseTextures   int
	MapTileBuilt     bool
	MapTileEntries   int
}

// GetIndexMemoryStats returns browse and map-tile index sizes.
func GetIndexMemoryStats() IndexMemoryStats {
	browseIndexMu.Lock()
	bm, bt := browseModels, browseTextures
	browseIndexMu.Unlock()

	mapTileIndexMu.Lock()
	mt := mapTileFileIndex
	mapTileIndexMu.Unlock()

	stats := IndexMemoryStats{
		BrowseBuilt:  bm != nil,
		MapTileBuilt: mt != nil,
	}
	if bm != nil {
		stats.BrowseModels = len(bm)
	}
	if bt != nil {
		stats.BrowseTextures = len(bt)
	}
	if mt != nil {
		stats.MapTileEntries = len(mt)
	}
	return stats
}

// AddEntry adds an entry to the listfile.
func AddEntry(fileDataID int, fileName string) {
	idLookup[fileDataID] = fileName
	nameLookup[fileName] = fileDataID
}

// ResetForCascUnload clears in-memory listfile state.
func ResetForCascUnload() {
	listfileLoaded = false
	isPreloaded = false
	preloadPromise = nil
	idLookup = map[int]string{}
	nameLookup = map[string]int{}
	preloadedIDLookup = map[int]string{}
	preloadedNameLookup = map[string]int{}
	resetFileIndexes()
}

var (
	browseM2WmoRegex   = regexp.MustCompile(`(?i)\.(m2|wmo)$`)
	browseBadWmoRegex  = regexp.MustCompile(`(?i)_([0-9]{3}|lod\d)\.wmo$`)
	browseTextureRegex = regexp.MustCompile(`(?i)\.(blp|png|tga|dds)$`)
	mapTileBlpRegex    = regexp.MustCompile(`^world/minimaps/([^/]+)/map(\d{1,2})_(\d{1,2})\.blp$`)
	mapTileAdtRegex    = regexp.MustCompile(`(?i)^world/maps/([^/]+)/([^/]+)_(\d{2})_(\d{2})\.adt$`)
)

var (
	browseIndexMu    sync.Mutex
	browseModels     []ListfileEntry
	browseTextures   []ListfileEntry
	mapTileIndexMu   sync.Mutex
	mapTileFileIndex []ListfileEntry
)

func resetFileIndexes() {
	browseIndexMu.Lock()
	browseModels = nil
	browseTextures = nil
	browseIndexMu.Unlock()

	mapTileIndexMu.Lock()
	mapTileFileIndex = nil
	mapTileIndexMu.Unlock()
}

func sortBrowseFileIndex(models, textures []ListfileEntry) {
	var wg sync.WaitGroup
	wg.Add(2)
	go func() {
		stringsort.SortBy(models, func(e ListfileEntry) string { return e.FileName })
		wg.Done()
	}()
	go func() {
		stringsort.SortBy(textures, func(e ListfileEntry) string { return e.FileName })
		wg.Done()
	}()
	wg.Wait()
}

func buildBrowseFileIndex() (models, textures []ListfileEntry) {
	models = make([]ListfileEntry, 0, 150000)
	textures = make([]ListfileEntry, 0, 800000)
	for fileDataID, fileName := range idLookup {
		if browseBadWmoRegex.MatchString(fileName) {
			continue
		}
		if browseM2WmoRegex.MatchString(fileName) {
			models = append(models, ListfileEntry{FileDataID: fileDataID, FileName: fileName})
		} else if browseTextureRegex.MatchString(fileName) {
			textures = append(textures, ListfileEntry{FileDataID: fileDataID, FileName: fileName})
		}
	}
	sortBrowseFileIndex(models, textures)
	return models, textures
}

// CollectBrowseFileIndex scans the loaded listfile once for browse model/texture entries.
func CollectBrowseFileIndex() (models, textures []ListfileEntry) {
	browseIndexMu.Lock()
	defer browseIndexMu.Unlock()
	if browseModels != nil {
		return browseModels, browseTextures
	}
	browseModels, browseTextures = buildBrowseFileIndex()
	return browseModels, browseTextures
}

func buildMapTileFileIndex() []ListfileEntry {
	entries := make([]ListfileEntry, 0, 50000)
	for fileDataID, fileName := range idLookup {
		lower := strings.ToLower(strings.ReplaceAll(fileName, `\`, `/`))
		if mapTileBlpRegex.MatchString(lower) || mapTileAdtRegex.MatchString(lower) {
			entries = append(entries, ListfileEntry{FileDataID: fileDataID, FileName: fileName})
		}
	}
	return entries
}

// CollectMapTileFileIndex scans the loaded listfile for minimap and ADT tile entries.
func CollectMapTileFileIndex() []ListfileEntry {
	mapTileIndexMu.Lock()
	defer mapTileIndexMu.Unlock()
	if mapTileFileIndex != nil {
		return mapTileFileIndex
	}
	mapTileFileIndex = buildMapTileFileIndex()
	return mapTileFileIndex
}
